ssstar/src/custom_credentials_provider.rs

```diff
@@ -0,0 +1,85 @@
+//! This module provides functionality of custom credentials provider
+//!
+//! The credentials are automatically refreshed if the credentials are expired that allows to use
+//! the this credential provider for long running jobs
+
+use aws_credential_types::provider::error::CredentialsError;
+use aws_credential_types::{provider::ProvideCredentials, Credentials};
+use futures::future::BoxFuture;
+use std::str::FromStr;
+use std::{fmt::Debug, sync::Arc};
+
+/// This represents the credentials to access to AWS API.
+#[derive(Clone, Debug)]
+pub struct CustomCredentialsUpdateOutput {
+    pub access_key: String,
+    pub secret_key: String,
+    pub session_token: Option<String>,
+}
+
+/// This is a trait that represents a callback to update credentials when they become expired
+pub trait CustomCredentialsUpdateCallback: Debug + Send + Sync {
+    /// This is not called for each AWS API call, this is called only when the
+    /// current credentials are expired
+    fn update_credentials(
+        &self,
+    ) -> BoxFuture<
+        '_,
+        std::result::Result<
+            CustomCredentialsUpdateOutput,
+            Box<dyn std::error::Error + Send + Sync + 'static>,
+        >,
+    >;
+}
+
+/// Custom credentials provider that actually implements `ProvideCredentials` and implements the needed
+/// traits to make it possible to put it into config as an option
+#[derive(Debug, Clone)]
+pub struct CustomCredentialsProvider(Arc<Box<dyn CustomCredentialsUpdateCallback>>);
```

Comment:
```suggestion
pub struct CustomCredentialsProvider(Arc<dyn CustomCredentialsUpdateCallback>);
```
`Arc<Box>` never makes sense. Just use only Arc

Reply:
you're right, probably I added `Box` when there' wasn `Arc`. Fixed